src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/ReferenceCountedDisposable.cs

```diff
@@ -270,20 +277,52 @@ public WeakReference(ReferenceCountedDisposable<T> reference)
             /// already been disposed.</returns>
             public ReferenceCountedDisposable<T>? TryAddReference()
             {
-                var weakInstance = _weakInstance;
-                if (weakInstance == null || !weakInstance.TryGetTarget(out var target))
+                // Ensure 'this' is only read once by this method.
+                var referenceCount = _boxedReferenceCount;
+                if (referenceCount == null)
                 {
+                    // This is either a default(WeakReference), or the current instance was constructed from a reference
+                    // that was already disposed. No target is available.
                     return null;
                 }
 
-                var referenceCount = _boxedReferenceCount;
-                if (referenceCount == null)
+                var weakInstance = referenceCount._weakInstance;
+
+                // _weakInstance is initialized by the constructor before assigning a value to _boxedReferenceCount.
```

Comment:
At first my thought was "what about write reordering?" but I guess I presume LazyInitialization.EnsureInitialized acts as a barrier.

Reply:
➡️ `LazyInitialization.EnsureInitialized` ensures the same value is returned to all callers. It tries to avoid calling the initialization delegate more than necessary, but does not ensure that it is called exactly once. I could do the same with a null check and `Interlocked.CompareExchange`, but this form is easier to understand the intent.